rust/azure_iot_operations_connector/src/data_processor/derived_json.rs

```diff
@@ -0,0 +1,704 @@
+            return Err(TransformErrorRepr::DuplicateField(output_field.to_string()));
+        }
+    }
+    // Create the output JSON from the `BTreeMap`
+    let output_json = serde_json::to_value(&output_btm)?;
+
+    // Derive the schema from the output JSON, removing the unnecessary examples metadata
+    let mut output_root_schema = schemars::schema_for_value!(&output_json);
+    if let Some(ref mut metadata) = output_root_schema.schema.metadata {
+        metadata.examples = vec![];
+    }
+
+    // Create a MessageSchema from the output JSON schema
+    let output_message_schema = MessageSchemaBuilder::default()
+        .content(serde_json::to_string(&output_root_schema)?)
+        .format(Format::JsonSchemaDraft07)
+        .schema_type(SchemaType::MessageSchema)
+        .build()?;
+
+    // Modify the input data struct to include the output JSON as the new payload
+    // NOTE: Because we are modifying the data in place, there should be NO ERRORS after this point.
+    data.payload = serde_json::to_vec(&output_json)?;
+
+    Ok(output_message_schema)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use azure_iot_operations_services::azure_device_registry::DatasetDataPoint;
+    use test_case::test_case;
+
+    struct TransformTestCase {
+        dataset: Dataset,
+        input_json: Value,
+        expected_output_json: Value,
+        expected_output_json_schema: Value,
+    }
+
+    /// Helper function to compare two `Data` structs for equality.
+    /// This is necessary over the PartialEq/Eq trait because when using JSON, we can
+    /// end up with different ordering of the keys in the JSON object, which prevents
+    /// us from being able to make accurate comparisons of the `payload` field.
+    fn json_data_eq(data1: &Data, data2: &Data) -> bool {
+        // Make new structs with the payload set to empty vectors to normalize our Data under
+        // comparison since we can't directly compare the payloads accurately.
+        let data1_no_payload = Data {
+            payload: Vec::new(),
+            ..data1.clone()
+        };
+        let data2_no_payload = Data {
+            payload: Vec::new(),
+            ..data2.clone()
+        };
+
+        // Deserialize the JSON payloads
+        let data1_json_payload: Value = serde_json::from_slice(&data1.payload).unwrap();
+        let data2_json_payload: Value = serde_json::from_slice(&data2.payload).unwrap();
+
+        // Now compare
+        data1_no_payload == data2_no_payload && data1_json_payload == data2_json_payload
+    }
+
+    /// Helper function to compare two `MessageSchema` structs for equality.
+    /// This is necessary over the PartialEq/Eq trait because when using JSON, we can
+    /// end up with different ordering of the keys in the JSON object, which prevents
+    /// us from being able to make accurate comparisons of the `content` field.
+    fn message_schmea_eq(schema1: &MessageSchema, schema2: &MessageSchema) -> bool {
```

Comment:
is this a fn that we should implement for MessageSchema? (maybe only for MessageSchemas of format JsonSchemaDraft07, but still)

Reply:
Maybe, although honestly I'm not convinced it's desirable. I think generally you would want normal `Eq`/`PartialEq`, just in this specific case for JSON we want to do it differently.

Reply:
we can leave this as an option to add later then - I think this might end up being a common MessageSchema issue since other implementations also would likely not have guaranteed order for their JSONs that they generate